File: python/exercises/code-smells/data-clumps/src/event_manager.py
"""
EventManager handles event creation and management

This class demonstrates the "Data Clumps" code smell by having multiple methods
that repeatedly use the same groups of parameters together:
- Date, time, and timezone parameters
- Address components (street, city, country, postal code)
- Coordinate parameters (latitude, longitude)
- Contact information (name, email, phone)

These parameter groups should be refactored into Parameter Objects.
"""

import math
import re
import uuid
from dataclasses import dataclass
from datetime import datetime

import pytz


@dataclass
class Event:
    """Represents an event with all its details"""

    id: str
    title: str
    description: str
    date: str
    start_time: str
    end_time: str
    timezone: str
    street: str
    city: str
    country: str
    postal_code: str
    latitude: float
    longitude: float
    organizer_name: str
    organizer_email: str
    organizer_phone: str
    max_attendees: int
    ticket_price: float
    attendees: int
    status: str


@dataclass
class Venue:
    """Represents a venue with location and contact details"""

    id: str
    name: str
    description: str
    street: str
    city: str
    country: str
    postal_code: str
    latitude: float
    longitude: float
    contact_name: str
    contact_email: str
    contact_phone: str
    capacity: int
    status: str


@dataclass
class Notification:
    """Represents a notification sent to an organizer"""

    id: str
    event_id: str
    organizer_name: str
    organizer_email: str
    organizer_phone: str
    message: str
    sent_at: str
    status: str


@dataclass
class TimeConversion:
    """Represents a time converted to a different timezone"""

    date: str
    time: str
    timezone: str


class EventManager:
    """Manages events, venues, and notifications with data clumps"""

    def __init__(self) -> None:
        self.events: dict[str, Event] = {}
        self.venues: dict[str, Venue] = {}
        self.notifications: dict[str, Notification] = {}

    def create_event(
        self,
        title: str,
        description: str,
        date: str,
        start_time: str,
        end_time: str,
        timezone_str: str,
        street: str,
        city: str,
        country: str,
        postal_code: str,
        latitude: float,
        longitude: float,
        organizer_name: str,
        organizer_email: str,
        organizer_phone: str,
        max_attendees: int = 100,
        ticket_price: float = 0.0,
    ) -> Event:
        """Creates a new event with date/time, location, and organizer information"""

        # Validate date and time
        if not self._is_valid_date(date):
            raise ValueError("Invalid date format")

        if not self._is_valid_time(start_time) or not self._is_valid_time(end_time):
            raise ValueError("Invalid time format")

        if not self._is_valid_timezone(timezone_str):
            raise ValueError("Invalid timezone")

        if not self._is_time_range_valid(start_time, end_time):
            raise ValueError("End time must be after start time")

        # Validate address
        if not self._is_valid_address(street, city, country, postal_code):
            raise ValueError("Invalid address information")

        # Validate coordinates
        if not self._is_valid_coordinates(latitude, longitude):
            raise ValueError("Invalid coordinates")

        # Validate contact information
        if not self._is_valid_contact(organizer_name, organizer_email, organizer_phone):
            raise ValueError("Invalid organizer contact information")

        event_id = self._generate_id()
        event = Event(
            id=event_id,
            title=title,
            description=description,
            date=date,
            start_time=start_time,
            end_time=end_time,
            timezone=timezone_str,
            street=street,
            city=city,
            country=country,
            postal_code=postal_code,
            latitude=latitude,
            longitude=longitude,
            organizer_name=organizer_name,
            organizer_email=organizer_email,
            organizer_phone=organizer_phone,
            max_attendees=max_attendees,
            ticket_price=ticket_price,
            attendees=0,
            status="active",
        )

        self.events[event_id] = event
        return event

    def update_event_timing(
        self,
        event_id: str,
        date: str,
        start_time: str,
        end_time: str,
        timezone_str: str,
    ) -> bool:
        """Updates event timing information"""

        if event_id not in self.events:
            return False

        # Validate date and time
        if not self._is_valid_date(date):
            raise ValueError("Invalid date format")

        if not self._is_valid_time(start_time) or not self._is_valid_time(end_time):
            raise ValueError("Invalid time format")

        if not self._is_valid_timezone(timezone_str):
            raise ValueError("Invalid timezone")

        if not self._is_time_range_valid(start_time, end_time):
            raise ValueError("End time must be after start time")

        event = self.events[event_id]
        event.date = date
        event.start_time = start_time
        event.end_time = end_time
        event.timezone = timezone_str

        return True

    def update_event_location(
        self,
        event_id: str,
        street: str,
        city: str,
        country: str,
        postal_code: str,
        latitude: float,
        longitude: float,
    ) -> bool:
        """Updates event location information"""

        if event_id not in self.events:
            return False

        # Validate address
        if not self._is_valid_address(street, city, country, postal_code):
            raise ValueError("Invalid address information")

        # Validate coordinates
        if not self._is_valid_coordinates(latitude, longitude):
            raise ValueError("Invalid coordinates")

        event = self.events[event_id]
        event.street = street
        event.city = city
        event.country = country
        event.postal_code = postal_code
        event.latitude = latitude
        event.longitude = longitude

        return True

    def register_venue(
        self,
        name: str,
        description: str,
        street: str,
        city: str,
        country: str,
        postal_code: str,
        latitude: float,
        longitude: float,
        contact_name: str,
        contact_email: str,
        contact_phone: str,
        capacity: int = 50,
    ) -> Venue:
        """Registers a new venue with address and contact information"""

        # Validate address
        if not self._is_valid_address(street, city, country, postal_code):
            raise ValueError("Invalid address information")

        # Validate coordinates
        if not self._is_valid_coordinates(latitude, longitude):
            raise ValueError("Invalid coordinates")

        # Validate contact information
        if not self._is_valid_contact(contact_name, contact_email, contact_phone):
            raise ValueError("Invalid contact information")

        venue_id = self._generate_id()
        venue = Venue(
            id=venue_id,
            name=name,
            description=description,
            street=street,
            city=city,
            country=country,
            postal_code=postal_code,
            latitude=latitude,
            longitude=longitude,
            contact_name=contact_name,
            contact_email=contact_email,
            contact_phone=contact_phone,
            capacity=capacity,
            status="active",
        )

        self.venues[venue_id] = venue
        return venue

    def send_event_notification(
        self,
        event_id: str,
        organizer_name: str,
        organizer_email: str,
        organizer_phone: str,
        message: str,
    ) -> bool:
        """Sends event notification to organizer"""

        if event_id not in self.events:
            return False

        # Validate contact information
        if not self._is_valid_contact(organizer_name, organizer_email, organizer_phone):
            raise ValueError("Invalid organizer contact information")

        notification_id = self._generate_id()
        notification = Notification(
            id=notification_id,
            event_id=event_id,
            organizer_name=organizer_name,
            organizer_email=organizer_email,
            organizer_phone=organizer_phone,
            message=message,
            sent_at=datetime.now().isoformat(),
            status="sent",
        )

        self.notifications[notification_id] = notification
        return True

    def schedule_recurring_event(
        self,
        base_event_id: str,
        pattern: str,
        occurrences: int,
        start_date: str,
        start_time: str,
        end_time: str,
        timezone_str: str,
    ) -> list[Event]:
        """Schedules recurring event"""

        if base_event_id not in self.events:
            raise ValueError("Base event not found")

        # Validate date and time
        if not self._is_valid_date(start_date):
            raise ValueError("Invalid date format")

        if not self._is_valid_time(start_time) or not self._is_valid_time(end_time):
            raise ValueError("Invalid time format")

        if not self._is_valid_timezone(timezone_str):
            raise ValueError("Invalid timezone")

        if not self._is_time_range_valid(start_time, end_time):
            raise ValueError("End time must be after start time")

        recurring_events = []
        base_event = self.events[base_event_id]

        for i in range(occurrences):
            event_date = self._calculate_next_date(start_date, pattern, i)
            event_id = self._generate_id()

            event = Event(
                id=event_id,
                title=base_event.title + " (Recurring)",
                description=base_event.description,
                date=event_date,
                start_time=start_time,
                end_time=end_time,
                timezone=timezone_str,
                street=base_event.street,
                city=base_event.city,
                country=base_event.country,
                postal_code=base_event.postal_code,
                latitude=base_event.latitude,
                longitude=base_event.longitude,
                organizer_name=base_event.organizer_name,
                organizer_email=base_event.organizer_email,
                organizer_phone=base_event.organizer_phone,
                max_attendees=base_event.max_attendees,
                ticket_price=base_event.ticket_price,
                attendees=0,
                status="active",
            )

            self.events[event_id] = event
            recurring_events.append(event)

        return recurring_events

    def calculate_distance(
        self, lat1: float, lon1: float, lat2: float, lon2: float
    ) -> float:
        """Calculates distance between two coordinate points"""

        if not self._is_valid_coordinates(lat1, lon1) or not self._is_valid_coordinates(
            lat2, lon2
        ):
            raise ValueError("Invalid coordinates")

        earth_radius = 6371  # Earth's radius in kilometers

        d_lat = math.radians(lat2 - lat1)
        d_lon = math.radians(lon2 - lon1)

        a = math.sin(d_lat / 2) * math.sin(d_lat / 2) + math.cos(
            math.radians(lat1)
        ) * math.cos(math.radians(lat2)) * math.sin(d_lon / 2) * math.sin(d_lon / 2)
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

        return earth_radius * c

    def find_events_in_date_range(
        self,
        start_date: str,
        start_time: str,
        end_date: str,
        end_time: str,
        timezone_str: str,
    ) -> list[Event]:
        """Finds events within date range"""

        # Validate date and time
        if not self._is_valid_date(start_date) or not self._is_valid_date(end_date):
            raise ValueError("Invalid date format")

        if not self._is_valid_time(start_time) or not self._is_valid_time(end_time):
            raise ValueError("Invalid time format")

        if not self._is_valid_timezone(timezone_str):
            raise ValueError("Invalid timezone")

        matching_events = []

        for event in self.events.values():
            if self._is_event_in_date_range(
                event, start_date, start_time, end_date, end_time, timezone_str
            ):
                matching_events.append(event)

        return matching_events

    def convert_time_to_timezone(
        self, date: str, time: str, from_timezone: str, to_timezone: str
    ) -> TimeConversion:
        """Converts time to different timezone"""

        if not self._is_valid_date(date):
            raise ValueError("Invalid date format")

        if not self._is_valid_time(time):
            raise ValueError("Invalid time format")

        if not self._is_valid_timezone(from_timezone) or not self._is_valid_timezone(
            to_timezone
        ):
            raise ValueError("Invalid timezone")

        # Create datetime in source timezone
        from_tz = pytz.timezone(from_timezone)
        to_tz = pytz.timezone(to_timezone)

        date_time_str = f"{date} {time}"
        naive_dt = datetime.strptime(date_time_str, "%Y-%m-%d %H:%M:%S")
        from_dt = from_tz.localize(naive_dt)
        to_dt = from_dt.astimezone(to_tz)

        return TimeConversion(
            date=to_dt.strftime("%Y-%m-%d"),
            time=to_dt.strftime("%H:%M:%S"),
            timezone=to_timezone,
        )

    # Validation methods (these contain the data clumps logic)
    def _is_valid_date(self, date: str) -> bool:
        """Validates date format"""
        if not re.match(r"^\d{4}-\d{2}-\d{2}$", date):
            return False
        try:
            datetime.strptime(date, "%Y-%m-%d")
            return True
        except ValueError:
            return False

    def _is_valid_time(self, time: str) -> bool:
        """Validates time format"""
        return bool(re.match(r"^\d{2}:\d{2}:\d{2}$", time))

    def _is_valid_timezone(self, timezone_str: str) -> bool:
        """Validates timezone"""
        try:
            pytz.timezone(timezone_str)
            return True
        except pytz.exceptions.UnknownTimeZoneError:
            return False

    def _is_time_range_valid(self, start_time: str, end_time: str) -> bool:
        """Validates that end time is after start time"""
        return start_time < end_time

    def _is_valid_address(
        self, street: str, city: str, country: str, postal_code: str
    ) -> bool:
        """Validates address components"""
        return (
            street.strip() != ""
            and city.strip() != ""
            and country.strip() != ""
            and postal_code.strip() != ""
        )

    def _is_valid_coordinates(self, latitude: float, longitude: float) -> bool:
        """Validates coordinate values"""
        return -90 <= latitude <= 90 and -180 <= longitude <= 180

    def _is_valid_contact(self, name: str, email: str, phone: str) -> bool:
        """Validates contact information"""
        email_pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
        return (
            name.strip() != ""
            and re.match(email_pattern, email) is not None
            and phone.strip() != ""
        )

    def _is_event_in_date_range(
        self,
        event: Event,
        start_date: str,
        start_time: str,
        end_date: str,
        end_time: str,
        timezone_str: str,
    ) -> bool:
        """Checks if event is within date range"""
        # Simplified comparison - in real implementation would handle timezone conversion
        event_datetime = f"{event.date} {event.start_time}"
        range_start = f"{start_date} {start_time}"
        range_end = f"{end_date} {end_time}"

        return range_start <= event_datetime <= range_end

    def _calculate_next_date(
        self, base_date: str, pattern: str, occurrence: int
    ) -> str:
        """Calculates next date based on pattern"""
        date = datetime.strptime(base_date, "%Y-%m-%d")

        if pattern == "daily":
            from datetime import timedelta

            date += timedelta(days=occurrence)
        elif pattern == "weekly":
            from datetime import timedelta

            date += timedelta(weeks=occurrence)
        elif pattern == "monthly":
            month = date.month + occurrence
            year = date.year + (month - 1) // 12
            month = ((month - 1) % 12) + 1
            date = date.replace(year=year, month=month)

        return date.strftime("%Y-%m-%d")

    def _generate_id(self) -> str:
        """Generates unique ID"""
        return f"evt_{uuid.uuid4().hex[:9]}"

    # Getters for testing
    def get_events(self) -> list[Event]:
        """Returns list of all events"""
        return list(self.events.values())

    def get_venues(self) -> list[Venue]:
        """Returns list of all venues"""
        return list(self.venues.values())

    def get_notifications(self) -> list[Notification]:
        """Returns list of all notifications"""
        return list(self.notifications.values())

File: typescript/exercises/code-smells/data-clumps/src/EventManager.ts
/**
 * EventManager handles event creation and management
 *
 * This class demonstrates the "Data Clumps" code smell by having multiple methods
 * that repeatedly use the same groups of parameters together:
 * - Date, time, and timezone parameters
 * - Address components (street, city, country, postal code)
 * - Coordinate parameters (latitude, longitude)
 * - Contact information (name, email, phone)
 *
 * These parameter groups should be refactored into Parameter Objects.
 */

interface Event {
  id: string;
  title: string;
  description: string;
  date: string;
  startTime: string;
  endTime: string;
  timezone: string;
  street: string;
  city: string;
  country: string;
  postalCode: string;
  latitude: number;
  longitude: number;
  organizerName: string;
  organizerEmail: string;
  organizerPhone: string;
  maxAttendees: number;
  ticketPrice: number;
  attendees: number;
  status: string;
}

interface Venue {
  id: string;
  name: string;
  description: string;
  street: string;
  city: string;
  country: string;
  postalCode: string;
  latitude: number;
  longitude: number;
  contactName: string;
  contactEmail: string;
  contactPhone: string;
  capacity: number;
  status: string;
}

interface Notification {
  id: string;
  eventId: string;
  organizerName: string;
  organizerEmail: string;
  organizerPhone: string;
  message: string;
  sentAt: string;
  status: string;
}

interface TimeConversion {
  date: string;
  time: string;
  timezone: string;
}

export class EventManager {
  private events: Map<string, Event> = new Map();
  private venues: Map<string, Venue> = new Map();
  private notifications: Map<string, Notification> = new Map();

  /**
   * Creates a new event with date/time, location, and organizer information
   */
  createEvent(
    title: string,
    description: string,
    date: string,
    startTime: string,
    endTime: string,
    timezone: string,
    street: string,
    city: string,
    country: string,
    postalCode: string,
    latitude: number,
    longitude: number,
    organizerName: string,
    organizerEmail: string,
    organizerPhone: string,
    maxAttendees: number = 100,
    ticketPrice: number = 0.0
  ): Event {
    // Validate date and time
    if (!this.isValidDate(date)) {
      throw new Error('Invalid date format');
    }

    if (!this.isValidTime(startTime) || !this.isValidTime(endTime)) {
      throw new Error('Invalid time format');
    }

    if (!this.isValidTimezone(timezone)) {
      throw new Error('Invalid timezone');
    }

    if (!this.isTimeRangeValid(startTime, endTime)) {
      throw new Error('End time must be after start time');
    }

    // Validate address
    if (!this.isValidAddress(street, city, country, postalCode)) {
      throw new Error('Invalid address information');
    }

    // Validate coordinates
    if (!this.isValidCoordinates(latitude, longitude)) {
      throw new Error('Invalid coordinates');
    }

    // Validate contact information
    if (!this.isValidContact(organizerName, organizerEmail, organizerPhone)) {
      throw new Error('Invalid organizer contact information');
    }

    const eventId = this.generateId();
    const event: Event = {
      id: eventId,
      title,
      description,
      date,
      startTime,
      endTime,
      timezone,
      street,
      city,
      country,
      postalCode,
      latitude,
      longitude,
      organizerName,
      organizerEmail,
      organizerPhone,
      maxAttendees,
      ticketPrice,
      attendees: 0,
      status: 'active',
    };

    this.events.set(eventId, event);
    return event;
  }

  /**
   * Updates event timing information
   */
  updateEventTiming(
    eventId: string,
    date: string,
    startTime: string,
    endTime: string,
    timezone: string
  ): boolean {
    const event = this.events.get(eventId);
    if (!event) {
      return false;
    }

    // Validate date and time
    if (!this.isValidDate(date)) {
      throw new Error('Invalid date format');
    }

    if (!this.isValidTime(startTime) || !this.isValidTime(endTime)) {
      throw new Error('Invalid time format');
    }

    if (!this.isValidTimezone(timezone)) {
      throw new Error('Invalid timezone');
    }

    if (!this.isTimeRangeValid(startTime, endTime)) {
      throw new Error('End time must be after start time');
    }

    event.date = date;
    event.startTime = startTime;
    event.endTime = endTime;
    event.timezone = timezone;

    return true;
  }

  /**
   * Updates event location information
   */
  updateEventLocation(
    eventId: string,
    street: string,
    city: string,
    country: string,
    postalCode: string,
    latitude: number,
    longitude: number
  ): boolean {
    const event = this.events.get(eventId);
    if (!event) {
      return false;
    }

    // Validate address
    if (!this.isValidAddress(street, city, country, postalCode)) {
      throw new Error('Invalid address information');
    }

    // Validate coordinates
    if (!this.isValidCoordinates(latitude, longitude)) {
      throw new Error('Invalid coordinates');
    }

    event.street = street;
    event.city = city;
    event.country = country;
    event.postalCode = postalCode;
    event.latitude = latitude;
    event.longitude = longitude;

    return true;
  }

  /**
   * Registers a new venue with address and contact information
   */
  registerVenue(
    name: string,
    description: string,
    street: string,
    city: string,
    country: string,
    postalCode: string,
    latitude: number,
    longitude: number,
    contactName: string,
    contactEmail: string,
    contactPhone: string,
    capacity: number = 50
  ): Venue {
    // Validate address
    if (!this.isValidAddress(street, city, country, postalCode)) {
      throw new Error('Invalid address information');
    }

    // Validate coordinates
    if (!this.isValidCoordinates(latitude, longitude)) {
      throw new Error('Invalid coordinates');
    }

    // Validate contact information
    if (!this.isValidContact(contactName, contactEmail, contactPhone)) {
      throw new Error('Invalid contact information');
    }

    const venueId = this.generateId();
    const venue: Venue = {
      id: venueId,
      name,
      description,
      street,
      city,
      country,
      postalCode,
      latitude,
      longitude,
      contactName,
      contactEmail,
      contactPhone,
      capacity,
      status: 'active',
    };

    this.venues.set(venueId, venue);
    return venue;
  }

  /**
   * Sends event notification to organizer
   */
  sendEventNotification(
    eventId: string,
    organizerName: string,
    organizerEmail: string,
    organizerPhone: string,
    message: string
  ): boolean {
    if (!this.events.has(eventId)) {
      return false;
    }

    // Validate contact information
    if (!this.isValidContact(organizerName, organizerEmail, organizerPhone)) {
      throw new Error('Invalid organizer contact information');
    }

    const notificationId = this.generateId();
    const notification: Notification = {
      id: notificationId,
      eventId,
      organizerName,
      organizerEmail,
      organizerPhone,
      message,
      sentAt: new Date().toISOString(),
      status: 'sent',
    };

    this.notifications.set(notificationId, notification);
    return true;
  }

  /**
   * Schedules recurring event
   */
  scheduleRecurringEvent(
    baseEventId: string,
    pattern: string,
    occurrences: number,
    startDate: string,
    startTime: string,
    endTime: string,
    timezone: string
  ): Event[] {
    const baseEvent = this.events.get(baseEventId);
    if (!baseEvent) {
      throw new Error('Base event not found');
    }

    // Validate date and time
    if (!this.isValidDate(startDate)) {
      throw new Error('Invalid date format');
    }

    if (!this.isValidTime(startTime) || !this.isValidTime(endTime)) {
      throw new Error('Invalid time format');
    }

    if (!this.isValidTimezone(timezone)) {
      throw new Error('Invalid timezone');
    }

    if (!this.isTimeRangeValid(startTime, endTime)) {
      throw new Error('End time must be after start time');
    }

    const recurringEvents: Event[] = [];

    for (let i = 0; i < occurrences; i++) {
      const eventDate = this.calculateNextDate(startDate, pattern, i);
      const eventId = this.generateId();

      const event: Event = {
        ...baseEvent,
        id: eventId,
        date: eventDate,
        startTime,
        endTime,
        timezone,
        title: baseEvent.title + ' (Recurring)',
        attendees: 0,
      };

      this.events.set(eventId, event);
      recurringEvents.push(event);
    }

    return recurringEvents;
  }

  /**
   * Calculates distance between two coordinate points
   */
  calculateDistance(
    lat1: number,
    lon1: number,
    lat2: number,
    lon2: number
  ): number {
    if (
      !this.isValidCoordinates(lat1, lon1) ||
      !this.isValidCoordinates(lat2, lon2)
    ) {
      throw new Error('Invalid coordinates');
    }

    const earthRadius = 6371; // Earth's radius in kilometers

    const dLat = this.degToRad(lat2 - lat1);
    const dLon = this.degToRad(lon2 - lon1);

    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this.degToRad(lat1)) *
        Math.cos(this.degToRad(lat2)) *
        Math.sin(dLon / 2) *
        Math.sin(dLon / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

    return earthRadius * c;
  }

  /**
   * Finds events within date range
   */
  findEventsInDateRange(
    startDate: string,
    startTime: string,
    endDate: string,
    endTime: string,
    timezone: string
  ): Event[] {
    // Validate date and time
    if (!this.isValidDate(startDate) || !this.isValidDate(endDate)) {
      throw new Error('Invalid date format');
    }

    if (!this.isValidTime(startTime) || !this.isValidTime(endTime)) {
      throw new Error('Invalid time format');
    }

    if (!this.isValidTimezone(timezone)) {
      throw new Error('Invalid timezone');
    }

    const matchingEvents: Event[] = [];

    for (const event of this.events.values()) {
      if (
        this.isEventInDateRange(event, startDate, startTime, endDate, endTime)
      ) {
        matchingEvents.push(event);
      }
    }

    return matchingEvents;
  }

  /**
   * Converts time to different timezone
   */
  convertTimeToTimezone(
    date: string,
    time: string,
    fromTimezone: string,
    toTimezone: string
  ): TimeConversion {
    if (!this.isValidDate(date)) {
      throw new Error('Invalid date format');
    }

    if (!this.isValidTime(time)) {
      throw new Error('Invalid time format');
    }

    if (
      !this.isValidTimezone(fromTimezone) ||
      !this.isValidTimezone(toTimezone)
    ) {
      throw new Error('Invalid timezone');
    }

    // Create date in source timezone
    const dateTime = new Date(`${date}T${time}`);

    // Convert to target timezone using Intl API
    const formatter = new Intl.DateTimeFormat('en-CA', {
      timeZone: toTimezone,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: false,
    });

    const parts = formatter.formatToParts(dateTime);
    const datePart = `${parts.find(p => p.type === 'year')?.value}-${parts.find(p => p.type === 'month')?.value}-${parts.find(p => p.type === 'day')?.value}`;
    const timePart = `${parts.find(p => p.type === 'hour')?.value}:${parts.find(p => p.type === 'minute')?.value}:${parts.find(p => p.type === 'second')?.value}`;

    return {
      date: datePart,
      time: timePart,
      timezone: toTimezone,
    };
  }

  // Validation methods (these contain the data clumps logic)
  private isValidDate(date: string): boolean {
    return /^\d{4}-\d{2}-\d{2}$/.test(date) && !isNaN(Date.parse(date));
  }

  private isValidTime(time: string): boolean {
    return /^\d{2}:\d{2}:\d{2}$/.test(time);
  }

  private isValidTimezone(timezone: string): boolean {
    try {
      Intl.DateTimeFormat(undefined, { timeZone: timezone });
      return true;
    } catch {
      return false;
    }
  }

  private isTimeRangeValid(startTime: string, endTime: string): boolean {
    return startTime < endTime;
  }

  private isValidAddress(
    street: string,
    city: string,
    country: string,
    postalCode: string
  ): boolean {
    return (
      street.trim() !== '' &&
      city.trim() !== '' &&
      country.trim() !== '' &&
      postalCode.trim() !== ''
    );
  }

  private isValidCoordinates(latitude: number, longitude: number): boolean {
    return (
      latitude >= -90 && latitude <= 90 && longitude >= -180 && longitude <= 180
    );
  }

  private isValidContact(name: string, email: string, phone: string): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return name.trim() !== '' && emailRegex.test(email) && phone.trim() !== '';
  }

  private isEventInDateRange(
    event: Event,
    startDate: string,
    startTime: string,
    endDate: string,
    endTime: string
  ): boolean {
    // Simplified comparison - in real implementation would handle timezone conversion
    const eventDateTime = `${event.date} ${event.startTime}`;
    const rangeStart = `${startDate} ${startTime}`;
    const rangeEnd = `${endDate} ${endTime}`;

    return eventDateTime >= rangeStart && eventDateTime <= rangeEnd;
  }

  private calculateNextDate(
    baseDate: string,
    pattern: string,
    occurrence: number
  ): string {
    const date = new Date(baseDate);

    switch (pattern) {
      case 'daily':
        date.setDate(date.getDate() + occurrence);
        break;
      case 'weekly':
        date.setDate(date.getDate() + occurrence * 7);
        break;
      case 'monthly':
        date.setMonth(date.getMonth() + occurrence);
        break;
    }

    return date.toISOString().split('T')[0]!;
  }

  private degToRad(deg: number): number {
    return deg * (Math.PI / 180);
  }

  private generateId(): string {
    return 'evt_' + Math.random().toString(36).substr(2, 9);
  }

  // Getters for testing
  getEvents(): Event[] {
    return Array.from(this.events.values());
  }

  getVenues(): Venue[] {
    return Array.from(this.venues.values());
  }

  getNotifications(): Notification[] {
    return Array.from(this.notifications.values());
  }
}
